File: client/src/loadCandidates.ts
/**
 * 候选加载器：从 candidates/<seam>/<候选>/meta.json 读出候选，按 seam+ref 索引。
 */
import { readFileSync, readdirSync, existsSync } from "node:fs";
import { join } from "node:path";
import { RegistryItem, L0Candidate, L1Signature } from "./contracts.js";

export interface CandidateMeta {
  registry_item: RegistryItem;
  l0: L0Candidate;
  l1: L1Signature;
  /** file-add 候选可为空对象；有接入口的给 import/register 片段 */
  barrel_snippet: { import?: string; register?: string };
  /** 候选根目录绝对路径，用于解析 files[].path */
  dir: string;
}

/** 加载某 candidates 根目录下所有候选，返回 seamId -> ref -> meta。 */
export function loadCandidates(candidatesRoot: string): Map<string, Map<string, CandidateMeta>> {
  const bySeam = new Map<string, Map<string, CandidateMeta>>();
  if (!existsSync(candidatesRoot)) return bySeam;

  for (const seamDir of readdirSync(candidatesRoot, { withFileTypes: true })) {
    if (!seamDir.isDirectory()) continue;
    const seamPath = join(candidatesRoot, seamDir.name);
    for (const candDir of readdirSync(seamPath, { withFileTypes: true })) {
      if (!candDir.isDirectory()) continue;
      const metaPath = join(seamPath, candDir.name, "meta.json");
      if (!existsSync(metaPath)) continue;

      const raw = JSON.parse(readFileSync(metaPath, "utf-8"));
      const meta: CandidateMeta = {
        registry_item: RegistryItem.parse(raw.registry_item),
        l0: L0Candidate.parse(raw.l0),
        l1: L1Signature.parse(raw.l1),
        barrel_snippet: raw.barrel_snippet ?? {},
        dir: join(seamPath, candDir.name),
      };
      const ref = meta.l0.ref;
      if (!bySeam.has(seamDir.name)) bySeam.set(seamDir.name, new Map());
      bySeam.get(seamDir.name)!.set(ref, meta);
    }
  }
  return bySeam;
}

/** 把候选清单铺平成 L0 列表（喂给选择引擎用）。 */
export function allL0(bySeam: Map<string, Map<string, CandidateMeta>>): L0Candidate[] {
  const out: L0Candidate[] = [];
  for (const refs of bySeam.values()) {
    for (const meta of refs.values()) out.push(meta.l0);
  }
  return out;
}

File: platform/load_candidates.py
"""候选加载器（Python 侧）：读 candidates/<seam>/<候选>/meta.json + files/。

与 client/src/loadCandidates.ts 同约定，供选择引擎使用。
"""

from __future__ import annotations

import json
from pathlib import Path

import loom_contracts as c


class Candidate:
    """一个候选的完整投影：L0/L1/RegistryItem + L2 文件全文（按需读）。"""

    def __init__(self, meta_path: Path):
        raw = json.loads(meta_path.read_text(encoding="utf-8"))
        self.dir = meta_path.parent
        self.registry_item = c.RegistryItem.model_validate(raw["registry_item"])
        self.l0 = c.L0Candidate.model_validate(raw["l0"])
        self.l1 = c.L1Signature.model_validate(raw["l1"])
        self.barrel_snippet: dict = raw.get("barrel_snippet", {})

    @property
    def ref(self) -> str:
        return self.l0.ref

    @property
    def seam_id(self) -> str:
        return self.l0.seam_id

    def l2_files(self) -> list[c.L2File]:
        """读出候选所有源码文件全文（L2 层，最贵，按需调用）。"""
        out: list[c.L2File] = []
        for f in self.registry_item.files:
            src = self.dir / f.path
            out.append(c.L2File(path=f.target, content=src.read_text(encoding="utf-8")))
        return out


def load_candidates(candidates_root: Path) -> dict[str, list[Candidate]]:
    """加载所有候选，返回 seam_id -> [Candidate]。"""
    by_seam: dict[str, list[Candidate]] = {}
    if not candidates_root.exists():
        return by_seam
    for seam_dir in sorted(candidates_root.iterdir()):
        if not seam_dir.is_dir():
            continue
        for cand_dir in sorted(seam_dir.iterdir()):
            meta = cand_dir / "meta.json"
            if not meta.exists():
                continue
            cand = Candidate(meta)
            by_seam.setdefault(cand.seam_id, []).append(cand)
    return by_seam


def find_candidate(by_seam: dict[str, list[Candidate]], seam_id: str, ref: str) -> Candidate | None:
    for cand in by_seam.get(seam_id, []):
        if cand.ref == ref:
            return cand
    return None
